File: teachprogramming/static/projects/stars.vb
Public Class Form1

    Dim background As Bitmap
    Dim r As Random = New Random()
    Dim stars(100) As Star

    Class Star
        Public xpos As Integer
        Public ypos As Integer
        Public speed As Integer
    End Class

    Private Sub Form1_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Load
        resize_background()
        Timer1.Start()
    End Sub

    Private Sub Timer1_Tick(ByVal sender As Object, ByVal e As System.EventArgs) Handles Timer1.Tick
        Graphics.FromImage(background).FillRectangle(Brushes.Black, 0, 0, PictureBox1.Width, PictureBox1.Height)
        For Each s As Star In stars
            s.xpos = s.xpos + s.speed
            If s.xpos > PictureBox1.Width Then
                s.xpos = 0
                s.ypos = r.Next(0, PictureBox1.Height - 1)
            End If
            Try
                background.SetPixel(s.xpos, s.ypos, Color.White)
            Catch ex As Exception
            End Try

        Next s
        PictureBox1.Refresh()
    End Sub

    Private Sub Form1_Resize(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Resize
        resize_background()
    End Sub

    Sub resize_background()
        PictureBox1.Width = Me.Width
        PictureBox1.Height = Me.Height
        background = New Bitmap(PictureBox1.Width, PictureBox1.Height)
        PictureBox1.Image = background

        For i As Integer = 0 To 100
            stars(i) = New Star
        Next
        For Each s As Star In stars
            s.xpos = r.Next(1, PictureBox1.Width - 1)
            s.ypos = r.Next(1, PictureBox1.Height - 1)
            s.speed = r.Next(1, 9)
        Next s
    End Sub
End Class

File: teachprogramming/static/projects/game/stars.cs

// Discussion about "Coupling" .. Do the Stars need to know about the rest of the system? how do they get this data?

namespace Stars

{


  class Star {

    public int x;

    public int y;

    public int speed;

    //public PictureBox pictureBox;


    public void move() { // int width

      x = x + speed;

      if (x >= Form1.width)

      { //pictureBox.Width

        x = 0;

      }

    }

  }


  public partial class Form1 : Form

  {

    int count = 0;

    Star[] stars = new Star[100];

    Random random = new Random();


    public static int width = 0;


    void log(string s) { System.Diagnostics.Debug.WriteLine(s); }

    public Form1()

    {

      InitializeComponent();

      timer1.Interval = 1000 / 60;

      log("hello");

      width = pictureBox1.Width;


      for (int i = 0; i < stars.Length; i++)

      {

        Star s = new Star();

        //s.pictureBox = pictureBox1;

        s.x = random.Next(0, pictureBox1.Width);

        s.y = random.Next(0, pictureBox1.Height);

        s.speed = random.Next(1,7);

        stars[i] = s;

      }


      timer1.Start();

    }


    private void timer1_Tick(object sender, EventArgs e)

    {

      log($"hello {count}");

      count += 1;


      Graphics g = pictureBox1.CreateGraphics();

      int width = (int)g.VisibleClipBounds.Width;

      int height = (int)g.VisibleClipBounds.Height;


      g.FillRectangle(new SolidBrush(Color.Black), 0, 0, width, height);

      g.FillRectangle(new SolidBrush(Color.White), count, 100, 20, 20);


      foreach (Star s in stars) {

        s.move();

        //if (s.x > width) {

        //  s.x = 0;

        //  s.y = random.Next(0, height);

        //}

        g.FillRectangle(new SolidBrush(Color.White), s.x, s.y, s.speed, 2);

      }

    }

  }

}

